Linking Arrow C++ fails on Windows

From the latest weekly verification run. First seen at https://github.com/apache/arrow-nanoarrow/actions/runs/4775468713/jobs/8489782873#step:12:200

```
   [ 44%] Linking CXX executable utils_test.exe
  C:/rtools43/x86_64-w64-mingw32.static.posix/bin/cmake.exe -E rm -f CMakeFiles/utils_test.dir/objects.a
  C:/rtools43/x86_64-w64-mingw32.static.posix/bin/ar.exe qc CMakeFiles/utils_test.dir/objects.a @CMakeFiles/utils_test.dir/objects1.rsp
  C:/rtools43/x86_64-w64-mingw32.static.posix/bin/c++.exe -Wl,--whole-archive CMakeFiles/utils_test.dir/objects.a -Wl,--no-whole-archive -o utils_test.exe -Wl,--out-implib,libutils_test.dll.a -Wl,--major-image-version,0,--minor-image-version,0 @CMakeFiles/utils_test.dir/linklibs.rsp
  C:\rtools43\x86_64-w64-mingw32.static.posix\bin/ld.exe: CMakeFiles/utils_test.dir/objects.a(utils_test.cc.obj):utils_test.cc:(.text+0x1443): undefined reference to `__imp__ZN5arrow18system_memory_poolEv'
  C:\rtools43\x86_64-w64-mingw32.static.posix\bin/ld.exe: CMakeFiles/utils_test.dir/objects.a(utils_test.cc.obj):utils_test.cc:(.text+0x145b): undefined reference to `__imp__ZN5arrow18system_memory_poolEv'
  C:\rtools43\x86_64-w64-mingw32.static.posix\bin/ld.exe: CMakeFiles/utils_test.dir/objects.a(utils_test.cc.obj):utils_test.cc:(.text+0x14a4): undefined reference to `__imp__ZN5arrow18system_memory_poolEv'
  C:\rtools43\x86_64-w64-mingw32.static.posix\bin/ld.exe: CMakeFiles/utils_test.dir/objects.a(utils_test.cc.obj):utils_test.cc:(.text+0x15e6): undefined reference to `__imp__ZN5arrow18system_memory_poolEv'
  C:\rtools43\x86_64-w64-mingw32.static.posix\bin/ld.exe: CMakeFiles/utils_test.dir/objects.a(utils_test.cc.obj):utils_test.cc:(.text+0x178e): undefined reference to `__imp__ZN5arrow18system_memory_poolEv'
  C:\rtools43\x86_64-w64-mingw32.static.posix\bin/ld.exe: CMakeFiles/utils_test.dir/objects.a(utils_test.cc.obj):utils_test.cc:(.text+0x206f): undefined reference to `__imp__ZN5arrow10Decimal12810FromStringEPKc'
  C:\rtools43\x86_64-w64-mingw32.static.posix\bin/ld.exe: CMakeFiles/utils_test.dir/objects.a(utils_test.cc.obj):utils_test.cc:(.text+0x20c4): undefined reference to `__imp__ZN5arrow10Decimal12810FromStringEPKc'
  C:\rtools43\x86_64-w64-mingw32.static.posix\bin/ld.exe: CMakeFiles/utils_test.dir/objects.a(utils_test.cc.obj):utils_test.cc:(.text+0x30e1): undefined reference to `__imp__ZN5arrow10Decimal25610FromStringEPKc'
  C:\rtools43\x86_64-w64-mingw32.static.posix\bin/ld.exe: CMakeFiles/utils_test.dir/objects.a(utils_test.cc.obj):utils_test.cc:(.text+0x3504): undefined reference to `__imp__ZN5arrow10Decimal25610FromStringEPKc'
  C:\rtools43\x86_64-w64-mingw32.static.posix\bin/ld.exe: CMakeFiles/utils_test.dir/objects.a(utils_test.cc.obj):utils_test.cc:(.text$_ZNO5arrow6ResultINS_10Decimal128EE10ValueOrDieEv[_ZNO5arrow6ResultINS_10Decimal128EE10ValueOrDieEv]+0x35): undefined reference to `__imp__ZN5arrow8internal17InvalidValueOrDieERKNS_6StatusE'
  C:\rtools43\x86_64-w64-mingw32.static.posix\bin/ld.exe: CMakeFiles/utils_test.dir/objects.a(utils_test.cc.obj):utils_test.cc:(.text$_ZNO5arrow6ResultINS_10Decimal256EE10ValueOrDieEv[_ZNO5arrow6ResultINS_10Decimal256EE10ValueOrDieEv]+0x35): undefined reference to `__imp__ZN5arrow8internal17InvalidValueOrDieERKNS_6StatusE'
```

--- a/.github/workflows/verify.yaml
+++ b/.github/workflows/verify.yaml
@@ -80,7 +80,7 @@ jobs:
         uses: actions/cache@v3
         with:
           path: arrow
-          key: arrow-${{ runner.os }}-2
+          key: arrow-${{ runner.os }}-3
 
       # For Arrow C++
       - name: Install Arrow C++ (MacOS)
@@ -103,10 +103,10 @@ jobs:
         if: matrix.config.label == 'windows' && steps.cache-arrow-build.outputs.cache-hit != 'true'
         shell: bash
         run: |
-          curl https://dlcdn.apache.org/arrow/arrow-11.0.0/apache-arrow-11.0.0.tar.gz | \
+          curl https://dlcdn.apache.org/arrow/arrow-11.0.0/apache-arrow-12.0.0.tar.gz | \
             tar -zxf -
           mkdir arrow-build && cd arrow-build
-          cmake ../apache-arrow-11.0.0/cpp -DCMAKE_INSTALL_PREFIX=../arrow
+          cmake ../apache-arrow-12.0.0/cpp -DCMAKE_INSTALL_PREFIX=../arrow
           cmake --build .
           cmake --install . --prefix=../arrow --config=Debug
           cd ..

--- a/.github/workflows/verify.yaml
+++ b/.github/workflows/verify.yaml
@@ -106,7 +106,7 @@ jobs:
           curl https://dlcdn.apache.org/arrow/arrow-12.0.0/apache-arrow-12.0.0.tar.gz | \
             tar -zxf -
           mkdir arrow-build && cd arrow-build
-          cmake ../apache-arrow-12.0.0/cpp -DCMAKE_INSTALL_PREFIX=../arrow
+          cmake ../apache-arrow-12.0.0/cpp -DCMAKE_INSTALL_PREFIX=../arrow -G 'Visual Studio 17 2022'
           cmake --build .
           cmake --install . --prefix=../arrow --config=Debug
           cd ..

--- a/.github/workflows/verify.yaml
+++ b/.github/workflows/verify.yaml
@@ -69,6 +69,8 @@ jobs:
       - uses: r-lib/actions/setup-r@v2
         with:
           use-public-rspm: true
+          windows-path-include-rtools: false
+          windows-path-include-mingw: false
       - uses: r-lib/actions/setup-r-dependencies@v2
         with:
           needs: check
@@ -103,10 +105,10 @@ jobs:
         if: matrix.config.label == 'windows' && steps.cache-arrow-build.outputs.cache-hit != 'true'
         shell: bash
         run: |
-          curl https://dlcdn.apache.org/arrow/arrow-12.0.0/apache-arrow-12.0.0.tar.gz | \
+          curl https://dlcdn.apache.org/arrow/arrow-11.0.0/apache-arrow-12.0.0.tar.gz | \
             tar -zxf -
           mkdir arrow-build && cd arrow-build
-          cmake ../apache-arrow-12.0.0/cpp -DCMAKE_INSTALL_PREFIX=../arrow -G 'Visual Studio 17 2022'
+          cmake ../apache-arrow-12.0.0/cpp -DCMAKE_INSTALL_PREFIX=../arrow
           cmake --build .
           cmake --install . --prefix=../arrow --config=Debug
           cd ..
@@ -115,7 +117,7 @@ jobs:
         if: matrix.config.label == 'windows'
         shell: bash
         run: |
-          echo "NANOARROW_CMAKE_OPTIONS=${NANOARROW_CMAKE_OPTIONS} -G 'Visual Studio 17 2022' -DArrow_DIR=$(pwd -W)/arrow/lib/cmake/Arrow -Dgtest_force_shared_crt=ON" >> $GITHUB_ENV
+          echo "NANOARROW_CMAKE_OPTIONS=${NANOARROW_CMAKE_OPTIONS} -DArrow_DIR=$(pwd -W)/arrow/lib/cmake/Arrow -Dgtest_force_shared_crt=ON" >> $GITHUB_ENV
 
       - name: Run dev/release/verify-release-candidate.sh
         shell: bash

--- a/.github/workflows/verify.yaml
+++ b/.github/workflows/verify.yaml
@@ -103,7 +103,7 @@ jobs:
         if: matrix.config.label == 'windows' && steps.cache-arrow-build.outputs.cache-hit != 'true'
         shell: bash
         run: |
-          curl https://dlcdn.apache.org/arrow/arrow-11.0.0/apache-arrow-12.0.0.tar.gz | \
+          curl https://dlcdn.apache.org/arrow/arrow-12.0.0/apache-arrow-12.0.0.tar.gz | \
             tar -zxf -
           mkdir arrow-build && cd arrow-build
           cmake ../apache-arrow-12.0.0/cpp -DCMAKE_INSTALL_PREFIX=../arrow

--- a/.github/workflows/verify.yaml
+++ b/.github/workflows/verify.yaml
@@ -115,7 +115,7 @@ jobs:
         if: matrix.config.label == 'windows'
         shell: bash
         run: |
-          echo "NANOARROW_CMAKE_OPTIONS=${NANOARROW_CMAKE_OPTIONS} -DArrow_DIR=$(pwd -W)/arrow/lib/cmake/Arrow -Dgtest_force_shared_crt=ON" >> $GITHUB_ENV
+          echo "NANOARROW_CMAKE_OPTIONS=${NANOARROW_CMAKE_OPTIONS} -G "Visual Studio 17 2022" -DArrow_DIR=$(pwd -W)/arrow/lib/cmake/Arrow -Dgtest_force_shared_crt=ON" >> $GITHUB_ENV
 
       - name: Run dev/release/verify-release-candidate.sh
         shell: bash

--- a/.github/workflows/verify.yaml
+++ b/.github/workflows/verify.yaml
@@ -115,7 +115,7 @@ jobs:
         if: matrix.config.label == 'windows'
         shell: bash
         run: |
-          echo "NANOARROW_CMAKE_OPTIONS=${NANOARROW_CMAKE_OPTIONS} -G "Visual Studio 17 2022" -DArrow_DIR=$(pwd -W)/arrow/lib/cmake/Arrow -Dgtest_force_shared_crt=ON" >> $GITHUB_ENV
+          echo "NANOARROW_CMAKE_OPTIONS=${NANOARROW_CMAKE_OPTIONS} -G 'Visual Studio 17 2022' -DArrow_DIR=$(pwd -W)/arrow/lib/cmake/Arrow -Dgtest_force_shared_crt=ON" >> $GITHUB_ENV
 
       - name: Run dev/release/verify-release-candidate.sh
         shell: bash

--- a/.github/workflows/verify.yaml
+++ b/.github/workflows/verify.yaml
@@ -105,7 +105,7 @@ jobs:
         if: matrix.config.label == 'windows' && steps.cache-arrow-build.outputs.cache-hit != 'true'
         shell: bash
         run: |
-          curl https://dlcdn.apache.org/arrow/arrow-11.0.0/apache-arrow-12.0.0.tar.gz | \
+          curl https://dlcdn.apache.org/arrow/arrow-12.0.0/apache-arrow-12.0.0.tar.gz | \
             tar -zxf -
           mkdir arrow-build && cd arrow-build
           cmake ../apache-arrow-12.0.0/cpp -DCMAKE_INSTALL_PREFIX=../arrow

